File: app/src/main/java/com/zygne/confetti/engine/explosions/BaseExplosion.java
package com.zygne.confetti.engine.explosions;

import com.zygne.confetti.engine.components.DynamicObject2d;
import com.zygne.confetti.engine.particles.BaseParticle;

/**
 * Created by Bardur Thomsen on 9/27/18.
 */
public abstract class BaseExplosion extends DynamicObject2d {

    public static final int STATE_ALIVE = 0;  // at least 1 particle is alive
    public static final int STATE_DEAD = 1;   // all particles are dead

    protected int state;
    protected final int numberOfParticles;
    protected float gravity;                    // the gravity of the explosion (+ upward, - down)
    protected float wind;                       // speed of wind on horizontal
    protected BaseParticle[] particles;         // particles in the explosion

    public BaseExplosion(float x, float y, int particleNr) {
        super(x, y);
        this.numberOfParticles = particleNr;
    }

    @Override
    public String toString() {
        return this.getClass().getSimpleName() + ": { x: " + this.position.getX() + ", y: " + this.position.getY() + ", particles: " + numberOfParticles +" }";
    }

    public float getGravity() {
        return gravity;
    }

    public void setGravity(float gravity) {
        this.gravity = gravity;
    }

    public float getWind() {
        return wind;
    }

    public void setWind(float wind) {
        this.wind = wind;
    }

    // helper methods -------------------------
    public boolean isAlive() {
        return this.state == STATE_ALIVE;
    }

    public boolean isDead() {
        return this.state == STATE_DEAD;
    }
}

File: app/src/main/java/com/zygne/confetti/engine/explosions/BaseExplosion.kt
package com.zygne.confetti.engine.explosions

import com.zygne.confetti.engine.components.DynamicObject2d
import com.zygne.confetti.engine.particles.BaseParticle
import com.zygne.confetti.engine.physics.Physics

abstract class BaseExplosion(x: Float, y: Float, protected val numberOfParticles: Int) : DynamicObject2d(x, y) {

    @JvmField
    var state = 0

    var gravity : Float = 0f // the gravity of the explosion (+ upward, - down) = 0f

    var wind  : Float = 0f // speed of wind on horizontal = 0f

    @JvmField
    var particles: Array<BaseParticle?> = emptyArray()

    @JvmField
    var physics: Physics = Physics()

    override fun toString(): String {
        return this.javaClass.simpleName + ": { x: " + position!!.x + ", y: " + position!!.y + ", particles: " + numberOfParticles + " }"
    }

    // helper methods -------------------------
    val isAlive: Boolean
        get() = state == STATE_ALIVE

    val isDead: Boolean
        get() = state == STATE_DEAD

    companion object {
        const val STATE_ALIVE = 0 // at least 1 particle is alive
        const val STATE_DEAD = 1 // all particles are dead
    }

}